File: app/src/main/java/dev/ragnarok/fenrir/model/Story.java
package dev.ragnarok.fenrir.model;

import android.os.Parcel;

import dev.ragnarok.fenrir.module.parcel.ParcelNative;

public class Story extends AbsModel implements ParcelNative.ParcelableNative {
    public static final Creator<Story> CREATOR = new Creator<Story>() {
        @Override
        public Story createFromParcel(Parcel in) {
            return new Story(in);
        }

        @Override
        public Story[] newArray(int size) {
            return new Story[size];
        }
    };
    public static final ParcelNative.Creator<Story> NativeCreator = Story::new;
    private int id;
    private int owner_id;
    private long date;
    private long expires_at;
    private boolean is_expired;
    private String access_key;
    private String target_url;
    private Photo photo;
    private Video video;
    private Owner author;

    public Story() {

    }

    protected Story(Parcel in) {
        super(in);
        id = in.readInt();
        owner_id = in.readInt();
        date = in.readLong();
        expires_at = in.readLong();
        is_expired = in.readInt() != 0;
        access_key = in.readString();
        target_url = in.readString();
        video = in.readParcelable(Video.class.getClassLoader());
        photo = in.readParcelable(Photo.class.getClassLoader());
        author = ParcelableOwnerWrapper.readOwner(in);
    }

    protected Story(ParcelNative in) {
        id = in.readInt();
        owner_id = in.readInt();
        date = in.readLong();
        expires_at = in.readLong();
        is_expired = in.readInt() != 0;
        access_key = in.readString();
        target_url = in.readString();
        video = in.readParcelable(Video.NativeCreator);
        photo = in.readParcelable(Photo.NativeCreator);
        author = ParcelableOwnerWrapper.readOwner(in);
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeInt(id);
        dest.writeInt(owner_id);
        dest.writeLong(date);
        dest.writeLong(expires_at);
        dest.writeInt(is_expired ? 1 : 0);
        dest.writeString(access_key);
        dest.writeString(target_url);
        dest.writeParcelable(video, flags);
        dest.writeParcelable(photo, flags);
        ParcelableOwnerWrapper.writeOwner(dest, flags, author);
    }

    @Override
    public void writeToParcelNative(ParcelNative dest) {
        dest.writeInt(id);
        dest.writeInt(owner_id);
        dest.writeLong(date);
        dest.writeLong(expires_at);
        dest.writeInt(is_expired ? 1 : 0);
        dest.writeString(access_key);
        dest.writeString(target_url);
        dest.writeParcelable(video);
        dest.writeParcelable(photo);
        ParcelableOwnerWrapper.writeOwner(dest, author);
    }

    public Photo getPhoto() {
        return photo;
    }

    public Story setPhoto(Photo photo) {
        this.photo = photo;
        return this;
    }

    public int getId() {
        return id;
    }

    public Story setId(int id) {
        this.id = id;
        return this;
    }

    public Video getVideo() {
        return video;
    }

    public Story setVideo(Video video) {
        this.video = video;
        return this;
    }

    public int getOwnerId() {
        return owner_id;
    }

    public Story setOwnerId(int ownerId) {
        owner_id = ownerId;
        return this;
    }

    public long getDate() {
        return date;
    }

    public Story setDate(long date) {
        this.date = date;
        return this;
    }

    public long getExpires() {
        return expires_at;
    }

    public Story setExpires(long expires_at) {
        this.expires_at = expires_at;
        return this;
    }

    public boolean isIs_expired() {
        return is_expired;
    }

    public Story setIs_expired(boolean is_expired) {
        this.is_expired = is_expired;
        return this;
    }

    public String getAccessKey() {
        return access_key;
    }

    public Story setAccessKey(String access_key) {
        this.access_key = access_key;
        return this;
    }

    public Owner getOwner() {
        return author;
    }

    public Story setOwner(Owner author) {
        this.author = author;
        return this;
    }

    public String getTarget_url() {
        return target_url;
    }

    public Story setTarget_url(String target_url) {
        this.target_url = target_url;
        return this;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Story))
            return false;

        Story story = (Story) o;
        return id == story.id && owner_id == story.owner_id;
    }

    @Override
    public int hashCode() {
        int result = id;
        result = 31 * result + owner_id;
        return result;
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/model/Story.kt
package dev.ragnarok.fenrir.model

import android.os.Parcel
import android.os.Parcelable
import dev.ragnarok.fenrir.getBoolean
import dev.ragnarok.fenrir.model.ParcelableOwnerWrapper.Companion.readOwner
import dev.ragnarok.fenrir.model.ParcelableOwnerWrapper.Companion.writeOwner
import dev.ragnarok.fenrir.module.parcel.ParcelNative
import dev.ragnarok.fenrir.putBoolean
import dev.ragnarok.fenrir.readTypedObjectCompat
import dev.ragnarok.fenrir.writeTypedObjectCompat

class Story : AbsModel, ParcelNative.ParcelableNative {
    var id = 0
        private set
    var ownerId = 0
        private set
    var date: Long = 0
        private set
    var expires: Long = 0
        private set
    var isIs_expired = false
        private set
    var accessKey: String? = null
        private set
    var target_url: String? = null
        private set
    var photo: Photo? = null
        private set
    var video: Video? = null
        private set
    var owner: Owner? = null
        private set

    constructor()
    internal constructor(`in`: Parcel) {
        id = `in`.readInt()
        ownerId = `in`.readInt()
        date = `in`.readLong()
        expires = `in`.readLong()
        isIs_expired = `in`.getBoolean()
        accessKey = `in`.readString()
        target_url = `in`.readString()
        video = `in`.readTypedObjectCompat(Video.CREATOR)
        photo = `in`.readTypedObjectCompat(Photo.CREATOR)
        owner = readOwner(`in`)
    }

    internal constructor(`in`: ParcelNative) {
        id = `in`.readInt()
        ownerId = `in`.readInt()
        date = `in`.readLong()
        expires = `in`.readLong()
        isIs_expired = `in`.readBoolean()
        accessKey = `in`.readString()
        target_url = `in`.readString()
        video = `in`.readParcelable(Video.NativeCreator)
        photo = `in`.readParcelable(Photo.NativeCreator)
        owner = readOwner(`in`)
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeInt(id)
        parcel.writeInt(ownerId)
        parcel.writeLong(date)
        parcel.writeLong(expires)
        parcel.putBoolean(isIs_expired)
        parcel.writeString(accessKey)
        parcel.writeString(target_url)
        parcel.writeTypedObjectCompat(video, flags)
        parcel.writeTypedObjectCompat(photo, flags)
        writeOwner(parcel, flags, owner)
    }

    override fun writeToParcelNative(dest: ParcelNative) {
        dest.writeInt(id)
        dest.writeInt(ownerId)
        dest.writeLong(date)
        dest.writeLong(expires)
        dest.writeBoolean(isIs_expired)
        dest.writeString(accessKey)
        dest.writeString(target_url)
        dest.writeParcelable(video)
        dest.writeParcelable(photo)
        writeOwner(dest, owner)
    }

    fun isEmptyStory(): Boolean {
        return photo == null && video == null && target_url.isNullOrEmpty()
    }

    @AbsModelType
    override fun getModelType(): Int {
        return AbsModelType.MODEL_STORY
    }

    fun setPhoto(photo: Photo?): Story {
        this.photo = photo
        return this
    }

    fun setId(id: Int): Story {
        this.id = id
        return this
    }

    fun setVideo(video: Video?): Story {
        this.video = video
        return this
    }

    fun setOwnerId(ownerId: Int): Story {
        this.ownerId = ownerId
        return this
    }

    fun setDate(date: Long): Story {
        this.date = date
        return this
    }

    fun setExpires(expires_at: Long): Story {
        expires = expires_at
        return this
    }

    fun setIs_expired(is_expired: Boolean): Story {
        isIs_expired = is_expired
        return this
    }

    fun setAccessKey(access_key: String?): Story {
        accessKey = access_key
        return this
    }

    fun setOwner(author: Owner?): Story {
        owner = author
        return this
    }

    fun setTarget_url(target_url: String?): Story {
        this.target_url = target_url
        return this
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun equals(other: Any?): Boolean {
        if (other !is Story) return false
        return id == other.id && ownerId == other.ownerId
    }

    override fun hashCode(): Int {
        var result = id
        result = 31 * result + ownerId
        return result
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<Story> = object : Parcelable.Creator<Story> {
            override fun createFromParcel(`in`: Parcel): Story {
                return Story(`in`)
            }

            override fun newArray(size: Int): Array<Story?> {
                return arrayOfNulls(size)
            }
        }
        val NativeCreator: ParcelNative.Creator<Story> =
            object : ParcelNative.Creator<Story> {
                override fun readFromParcelNative(dest: ParcelNative): Story {
                    return Story(dest)
                }
            }
    }
}